### PR TITLE
[JUJU-121] Adds support for auto create of instance profiles.

### PR DESCRIPTION
**Pull request by tlm**
This commit adds support instance profiles auto creation. Including the associated role and policy needed to scope permissions correctly.

Accompanying this PR is also a discourse doc on the permissions Juju requires when talking with AWS.

## QA steps

```sh
juju bootstrap --bootstrap-constraints="instance-role=auto" aws/ap-southeast-2 test-tlm-controller
```

You then need to confirm with the aws cli that the associated instance profile elements were created. Make sure that you see a role with the same here attached to the instance profile.

```sh
aws iam get-instance-profile --instance-profile-name juju-controller-test-tlm-controller
```

Add some machines to Juju and deploy a charm with storage to confirm permissions are operating correctly.

Check HA works

```
$ juju enable-ha
juju enable-ha
maintaining machines: 0
adding machines: 1, 2
$ juju status
Model       Controller     Cloud/Region        Version   SLA          Timestamp
controller  test-tlm-controller aws/ap-southeast-2  2.9.20.1  unsupported  09:34:28+10:00

Machine  State    DNS           Inst id              Series  AZ               Message
0        started  13.211.59.65  i-f00f00f00f00f00f00  focal   ap-southeast-2b  running
1        pending                pending              focal                    attaching aws instance profile arn:aws:iam::123456789:instance-profile/juju-controller-test-tlm-controller
2        pending                pending              focal                    attaching aws instance profile arn:aws:iam::123456789:instance-profile/juju-controller-test-tlm-controller
```

Check destroy-controller works
```
$ juju destroy-controller test-tlm-controller --destroy-all-models
```

## Documentation changes

Permission list: https://discourse.charmhub.io/t/juju-aws-permissions/5307
Doc Link: https://discourse.charmhub.io/t/using-aws-instance-profiles-with-juju-2-9/5185/3

## Bug reference

N/A

### CODE REVIEW

File: provider/ec2/environ.go [hunk 1]
```diff
@@ -214,31 +214,23 @@ func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.Provi
...
-		//instProfile, err := ensureControllerInstanceProfile(
-		//	ctx.Context(),
-		//	e.iamClient,
-		//	controllerName,
-		//	controllerUUID)
-		//if err != nil {
-		//	return nil, err
-		//}
-		//args.BootstrapConstraints.InstanceRole = instProfile.InstanceProfileName
+		controllerName, ok := args.ControllerConfig[controller.ControllerName].(string)
```
Comment (hpidcock) [inline provider/ec2/environ.go:217]: This feels like it should actually check the map access not the type assertion, same below.

Comment (hpidcock) [reply provider/ec2/environ.go:217]: well maybe even check the type assertion too

File: provider/ec2/iam.go [hunk 2]
```diff
@@ -112,12 +128,98 @@ func ensureControllerInstanceProfile(
 		var alreadyExistsErr *iamtypes.EntityAlreadyExistsException
 		if stderrors.As(err, &alreadyExistsErr) {
 			// Instance Profile already exists so we don't need todo anything. Let just find it
-			return findInstanceProfileFromName(ctx, client, profileName)
+			ip, e := findInstanceProfileFromName(ctx, client, profileName)
```
Comment (hpidcock) [inline provider/ec2/iam.go:131]: e can just be err here since we already handled it.

File: provider/ec2/iam.go [hunk 3]
```diff
@@ -112,12 +128,98 @@ func ensureControllerInstanceProfile(
...
+				Value: aws.String(controllerUUID),
+			},
+		},
+	})
+
+	if err != nil {
+		var alreadyExistsErr *iamtypes.EntityAlreadyExistsException
+		if stderrors.As(err, &alreadyExistsErr) {
+			// Role already exists so we don't need todo anything. Let just find it
+			r, e := findRoleFromName(ctx, client, roleName)
```
Comment (hpidcock) [inline provider/ec2/iam.go:192]: same here

File: provider/ec2/iam_docs.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	  },
+	  "Action": "sts:AssumeRole"
+	}
+  ]
+}
+`
+	// controllerRolePolicy is the AWS IAM policy used for controller role
+	// permissions. This JSON document must be kept in line with the AWS
+	// permissions used by Juju.
+	controllerRolePolicy = `}
```
Comment (hpidcock) [inline provider/ec2/iam_docs.go:23]: `}`?

File: provider/ec2/iam_docs.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
+package ec2
+
+const (
+	// controllerRoleAssumePolicy describes the polciy for the controller roll
+	// stating what principals can assume the role. We only allow ec2 instances
+	// in this case.
+	controllerRoleAssumePolicy = ` {
```
Comment (hpidcock) [inline provider/ec2/iam_docs.go:7]: leading space?

**Review by hpidcock — COMMENTED** [review]
Only three issues I could see
- `juju enable-ha` spins up controllers without an IAM profile. 
- Teardown isn't working
```
machine-0: 12:58:37 ERROR juju.worker.dependency "undertaker" manifold worker returned unexpected error: cannot destroy cloud resources: destroying storage: listing volumes: operation error EC2: DescribeVolumes, https response error StatusCode: 403, RequestID: eebe4320-5260-448c-91d1-9b74fe7db8ac, api error UnauthorizedOperation: You are not authorized to perform this operation.
```
- Instance profile is partially left in bootstrap status (I have a fix for this)

File: provider/ec2/iam.go [hunk 6]
```diff
@@ -112,12 +122,98 @@ func ensureControllerInstanceProfile(
...
+			return ip, cleanups, err
 		}
 		// Some other error that we can't recover from.
-		return nil, errors.Annotate(err, "creating controller instance profile")
+		return nil, cleanups, errors.Annotate(err, "creating controller instance profile")
 	}
-	return res.InstanceProfile, nil
+
+	cleanups = append([]func(){func() {
+		_, _ = client.DeleteInstanceProfile(ctx, &iam.DeleteInstanceProfileInput{
```
Comment (hpidcock) [inline provider/ec2/iam.go:133]: at least log out the error if one is returned

File: provider/ec2/iam.go [hunk 7]
```diff
@@ -112,12 +122,98 @@ func ensureControllerInstanceProfile(
...
+		return nil, cleanups, errors.Annotatef(
+			err,
+			"attaching role %s to instance profile %s",
+			*role.RoleName,
+			*res.InstanceProfile.InstanceProfileName,
+		)
+	}
+
+	cleanups = append([]func(){func() {
+		_, _ = client.RemoveRoleFromInstanceProfile(ctx, &iam.RemoveRoleFromInstanceProfileInput{
```
Comment (hpidcock) [inline provider/ec2/iam.go:153]: same here

File: provider/ec2/iam.go [hunk 8]
```diff
@@ -112,12 +122,98 @@ func ensureControllerInstanceProfile(
...
+			// Role already exists so we don't need todo anything. Let just find it
+			r, err := findRoleFromName(ctx, client, roleName)
+			return r, cleanups, err
+		}
+		// Some other error that we can't recover from.
+		return nil, cleanups, errors.Annotate(err, "creating controller instance role")
+	}
+
+	cleanups = append(cleanups, func() {
+		_, _ = client.DeleteRole(ctx, &iam.DeleteRoleInput{
```
Comment (hpidcock) [inline provider/ec2/iam.go:194]: same here

File: provider/ec2/iam.go [hunk 9]
```diff
@@ -112,12 +122,98 @@ func ensureControllerInstanceProfile(
...
+		PolicyName:     aws.String(roleName),
+		RoleName:       res.Role.RoleName,
+	})
+
+	if err != nil {
+		return nil, cleanups, errors.Annotatef(err, "attaching role %s policy", *res.Role.RoleName)
+	}
+
+	cleanups = append([]func(){func() {
+		_, _ = client.DeleteRolePolicy(ctx, &iam.DeleteRolePolicyInput{
```
Comment (hpidcock) [inline provider/ec2/iam.go:210]: same here

**Review by hpidcock — COMMENTED** [review]
Still testing will approve as soon as I'm done 🏖️

**Review by hpidcock — APPROVED** [review]
LGTM. Conditional on the cleanup errors are logged.

**Comment by hpidcock** [comment]
@tlm lets add a backlog item to deal with instance profile cleanup in github.com/provider/ec2.destroyControllerManagedModels

**Comment by tlm** [comment]
$$merge$$

**Pull request merged**